src/java.base/share/classes/java/util/WeakHashMap.java

```diff
@@ -283,8 +283,14 @@ static Object unmaskNull(Object key) {
      * Checks for equality of non-null reference x and possibly-null y.  By
      * default uses Object.equals.
      */
-    private static boolean eq(Object x, Object y) {
-        return x == y || x.equals(y);
+    private boolean matchesKey(Entry<K,V> e, Object key) {
+        // check if the given entry refers to the given key without
+        // keeping a strong reference to the entry's referent
+        if (e.refersTo(key)) return true;
+
+        // then checks for equality
+        Object k = e.get();
+        return key == k || key.equals(k);
```

Comment:
I think `key == k` is already covered by refersTo. But k could be null; checking for that here might be useful, to skip the call to equals in that case.

Reply:
Good point on checking k != null. A cleaner check:

```
        // then check for equality if the referent is not cleared
        Object k = e.get();
        return k != null || key.equals(k);
```

Reply:
You meant: `return k != null && key.equals(k);` right?

Reply:
oops...yes.